google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java

```diff
@@ -288,6 +288,34 @@ public boolean next() throws SpannerException {
           }
         }
 
+        @Override
+        public boolean initiateStreaming(
+            AsyncResultSet.StreamMessageListener streamMessageListener) {
+          try {
+            boolean streamInitiated = super.initiateStreaming(streamMessageListener);
+            if (!streamInitiated) {
+              synchronized (lock) {
+                session.get().markUsed();
+                sessionUsedForQuery = true;
+              }
+            }
+            if (!streamInitiated && isSingleUse) {
+              close();
```

Comment:
I struggle to understand why we do this. If no streaming was initiated, that means that the result set does not support streaming. But I would not expect that we should then just close it.

It might be a bit of a theoretical discussion, as at the moment, the only result set implementation that we have is the streaming gRPC result set, but in theory we could in the future also add support for running queries with the `ExecuteSql` RPC (instead of `ExecuteStreamingSql`). That would then not support streaming, and would then end up being closed here.